src/app/components/chat/window/output/AppChatWindowOutputItem.vue

```diff
@@ -0,0 +1,530 @@
+	if (canRemoveMessage.value) {
+		++itemCount;
+	}
+	if (canEditMessage.value) {
+		++itemCount;
+	}
+
+	return itemWidth * itemCount + (itemCount - 1);
+});
+
+const canRemoveMessage = computed(() => {
+	if (!chat.value.currentUser) {
+		return false;
+	}
+
+	// The owner of the message can remove it.
+	if (chat.value.currentUser.id === message.value.user.id) {
+		return true;
+	}
+
+	// Mods/Room owners can also remove the message.
+	return userCanModerateOtherUser(
+		chat.value,
+		room.value,
+		chat.value.currentUser,
+		message.value.user
+	);
+});
+
+const canEditMessage = computed(() => {
+	// Only content messages can be edited.
+	if (message.value.type !== 'content') {
+		return false;
+	}
+
+	if (!chat.value.currentUser) {
+		return false;
+	}
+
+	// Only the owner of the message can edit.
+	return chat.value.currentUser.id === message.value.user.id;
+});
+
+function startEdit() {
+	setMessageEditing(chat.value, message.value);
+	Popper.hideAll();
+}
+
+function stopEdit() {
+	setMessageEditing(chat.value, null);
+	Popper.hideAll();
+}
+
+async function removeMessage() {
+	Popper.hideAll();
+
+	const result = await ModalConfirm.show(
+		$gettext(`Are you sure you want to remove this message?`)
+	);
+
+	if (!result) {
+		return;
+	}
+
+	setMessageEditing(chat.value, null);
+	chatRemoveMessage(chat.value, room.value, message.value.id);
+}
+
+let canClearFocus = false;
+let isFocused = false;
+
+async function onFocusItem() {
+	isFocused = true;
+	// Allow a second click to blur only after ~1/2 of our animations play.
+	setTimeout(() => (canClearFocus = isFocused), 100);
+}
+
+function onBlurItem() {
+	isFocused = false;
+	canClearFocus = false;
+}
+
+async function onClickItem() {
+	if (isFocused && canClearFocus) {
+		itemWrapper.value?.blur();
+	}
+}
+
+function onItemWrapperResize() {
+	if (!itemWrapper.value) {
+		return;
+	}
+
+	const { offsetWidth } = itemWrapper.value;
+	const decoratorPadding = 12 * 2;
+	contentViewerBounds.width = Math.max(offsetWidth - decoratorPadding, 100);
+}
+</script>
+
+<template>
+	<div
+		ref="root"
+		class="chat-window-output-item"
+		:class="{
+			'-message-queued': message._showAsQueued,
+			'-message-editing': isEditing,
+		}"
+	>
+		<a v-if="message.showAvatar" class="-avatar">
+			<AppPopper placement="right">
+				<img
+					class="-avatar-img img-responsive"
+					:src="message.user.img_avatar"
+					alt=""
+					draggable="false"
+				/>
+
+				<template #popover>
+					<AppChatUserPopover :user="message.user" :room="room" />
+				</template>
+			</AppPopper>
+		</a>
+
+		<div
+			ref="itemWrapper"
+			v-app-observe-dimensions="onItemWrapperResize"
+			class="-item-container-wrapper"
+			tabindex="-1"
+			@focus="onFocusItem"
+			@blur="onBlurItem"
+			@click.capture="onClickItem"
+		>
+			<div class="-item-container">
+				<div
+					class="-item-decorator"
+					:style="{
+						'background-image': isEditingColor
+							? `linear-gradient(${isEditingColor}, ${isEditingColor}`
```

Comment:
why are we using a linear gradient? so that we don't need to do an absolute element that expands across the whole thing?
Also, there's a missing `)` at the end of this line I think

Reply:
replaced with a border instead